File: strategy.py
import random
from engine import Game, KNOWN_WORDS, compare_words
from collections import defaultdict
import statistics


def splits(guess, candidates):
  """ sizes of groups that yield the same pattern """
  groups = defaultdict(int)
  
  for candidate in candidates:
    result = compare_words(guess, candidate)
    groups[result] += 1
    
  return sorted(groups.values())


def best_candidate(candidates):
  if len(candidates) == 1:
    return candidates[0]
      
  best_guess = None
  best_score = float("inf")
  
  list = KNOWN_WORDS if len(candidates) > 10 else candidates
  
  for guess in list:
    score = statistics.geometric_mean(splits(guess, candidates))
    
    if score < best_score:
      best_score = score
      best_guess = guess
      
  return best_guess


def play(game: Game) -> str:
  """
  Plays the game and finds the secret word.
  """
  candidates = KNOWN_WORDS
  
  while True:
    guess = best_candidate(candidates)
    result_pattern = game.make_a_guess(guess)
    
    if result_pattern == "🟩🟩🟩🟩🟩":
      return guess
    
    candidates = [word for word in candidates if compare_words(guess, word) == result_pattern]

File: strategy.mjs
import { KNOWN_WORDS, compareWords } from './engine.mjs';
import { geometricMean } from 'simple-statistics';
import { default as _ } from 'lodash';

function splits(guess, candidates) {
  const groups = _.countBy(candidates, candidate => compareWords(guess, candidate));
  return Object.values(groups).sort((a, b) => a - b);
}

function bestCandidate(candidates) {
  if (candidates.length === 1) return candidates[0];

  let bestGuess = null;
  let bestScore = Infinity;

  const list = candidates.length > 10 ? KNOWN_WORDS : candidates;

  let idx = 0;

  for (const guess of list) {
    const score = geometricMean(splits(guess, candidates));

    if (score < bestScore) {
      bestScore = score;
      bestGuess = guess;
    }

    if (++idx % 100 === 0) {
        console.log(`Processed ${idx} words`);
    }
  }

  return bestGuess;
}

export default function play(game) {
  let candidates = KNOWN_WORDS;

  while (true) {
    const guess = bestCandidate(candidates);
    const resultPattern = game.makeAGuess(guess);

    if (resultPattern === "🟩🟩🟩🟩🟩") return guess;

    candidates = candidates.filter(word => compareWords(guess, word) === resultPattern);
  }
}
